gitpod-shared/src/telemetryService.ts

```diff
@@ -3,59 +3,324 @@
+import { Disposable } from './common/dispose';
+import { cloneAndChange, escapeRegExpCharacters, mixin } from './common/utils';
+import fetch from 'node-fetch';
+
+const ProductionUntrustedSegmentKey = 'untrusted-dummy-key';
+
+export interface TelemetryEventProperties {
+	[key: string]: any;
+}
+
+export interface UserFlowTelemetryProperties {
+	flow: string;
+
+	gitpodVersion?: string;
+
+	workspaceId?: string;
+	instanceId?: string;
+
+	userId?: string;
+
+	[key: string]: any;
+}
+
+export interface ITelemetryService {
+	sendTelemetryEvent(eventName: string, properties?: TelemetryEventProperties): void;
+	sendTelemetryException(error: Error, properties?: TelemetryEventProperties): void;
+}
+
+const TRUSTED_VALUES = new Set([
+	'gitpodHost'
+]);
+
+export class TelemetryService extends Disposable implements ITelemetryService {
+	private analitycsClients: Map<string, Analytics> = new Map();
+	private telemetryLogger: vscode.TelemetryLogger;
+
+	constructor(extensionId: string, extensionVersion: string, segmentKey: string, piiPaths: string[], userId: string, gitpodHost: string, private readonly logService: ILogService) {
+		super();
+
+		// static cleanup pattern for: `vscode-file:///DANGEROUS/PATH/resources/app/Useful/Information`
+		const cleanupPatterns = [/(vscode-)?file:\/\/\/.*?\/resources\/app\//gi];
+
+		piiPaths.push(vscode.env.appRoot);
+		piiPaths.push(os.tmpdir());
+		piiPaths.push(os.homedir());
+		for (const piiPath of piiPaths) {
+			cleanupPatterns.push(new RegExp(escapeRegExpCharacters(piiPath), 'gi'));
+
+			if (piiPath.indexOf('\\') >= 0) {
+				cleanupPatterns.push(new RegExp(escapeRegExpCharacters(piiPath.replace(/\\/g, '/')), 'gi'));
+			}
+		}
+
+		const commonProperties = getCommonProperties(extensionId, extensionVersion);
 
-type RequireKeys<T extends object, K extends keyof T> =
-	Required<Pick<T, K>> & Omit<T, K>;
-
-export type TelemetrySettings = RequireKeys<AnalyticsSettings, 'writeKey' | 'host' | 'path'>
-
-const analyticsClientFactory = async (pendingUserId: Promise<string>, settings: TelemetrySettings, logger: ILogService): Promise<BaseTelemetryClient> => {
-	const userId = await pendingUserId;
-	let segmentAnalyticsClient = new Analytics(settings);
-	logger.debug("analytics: " + new URL(settings.path, settings.host).href.replace(/\/$/, '')); // aligned with how segment does it internally
-
-	// Sets the analytics client into a standardized form
-	const telemetryClient: BaseTelemetryClient = {
-		logEvent: (eventName: string, data?: AppenderData) => {
-			try {
-				segmentAnalyticsClient.track({
-					userId,
-					event: eventName,
-					properties: data?.properties
-				}, (e) => {
-					if (e) {
-						logger.error('Failed to log event to app analytics:', e);
+		this.telemetryLogger = this._register(vscode.env.createTelemetryLogger(
+			{
+				sendEventData: (eventName, data) => {
+					const idx = eventName.indexOf('/');
+					eventName = eventName.substring(idx + 1);
+
+					const properties = data ?? {};
```

Comment:
```suggestion
					const properties = mixin(data ?? {}, commonProperties);
```
common properties are missing

Reply:
see
<img width="1203" alt="image" src="https://github.com/gitpod-io/gitpod-code/assets/20944364/3f263134-2db6-436c-bf95-91f352083bdd">

Reply:
they should be already there as they are added internally by vscode , that's should only needed for `sendErrorData`  🤔

Reply:
same in gitpod-vscode-desktop https://github.com/gitpod-io/gitpod-vscode-desktop/blob/1b92e4a3809a339689e9740f0b4313b1454219f4/src/common/telemetry.ts#L72

Reply:
But you see the screenshot, it [fixed](https://github.com/gitpod-io/gitpod-code/pull/11/commits/a8156410ef47d6c555f090f28093ea5ccb0381c2) (version / name are correct) after I mix the common properties

Reply:
So the common properties are missing in `additionalCommonProperties` ?

Reply:
ah wait you tested from sources? then that's expected, it's because vscode will do validation for files in extension folder and telemetryService is in a different folder `gitpod-shared`, we should test it using webpacked version, e.g. in preview env using updaed image tag